### PR TITLE
podman pod stats

### PR DESCRIPTION
**Pull request by baude**
add the ability to monitor container statistics in a pod.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@baude looks like you've a merge conflict to work through.

File: cmd/podman/pod_stats.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	defer runtime.Shutdown(false)
+
+	times := -1
+	if c.Bool("no-stream") {
+		times = 1
+	}
+
+	if len(c.Args()) > 0 {
+		podFunc = func() ([]*libpod.Pod, error) { return runtime.GetPodsByList(c.Args()) }
+	} else if c.IsSet("all") {
```
Comment (haircommander) [inline cmd/podman/pod_stats.go]: Should this be the local variable 'all'?

File: cmd/podman/pod_stats.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+			Usage: "show stats for all pods.  Only running pods are shown by default.",
+		},
+		cli.BoolFlag{
+			Name:  "no-stream",
+			Usage: "disable streaming stats and only pull the first result, default setting is false",
+		},
+		cli.BoolFlag{
+			Name:  "no-reset",
+			Usage: "disable resetting the screen between intervals",
+		}, LatestFlag,
```
Comment (mheon) [inline cmd/podman/pod_stats.go]: Move LatestFlag to the end, on its own line

File: cmd/podman/pod_stats.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	times := -1
+	if c.Bool("no-stream") {
+		times = 1
+	}
+
+	if len(c.Args()) > 0 {
+		podFunc = func() ([]*libpod.Pod, error) { return runtime.GetPodsByList(c.Args()) }
+	} else if c.IsSet("all") {
+		podFunc = runtime.GetAllPods
+	} else {
```
Comment (haircommander) [inline cmd/podman/pod_stats.go]: Is latest going to be an option?

File: libpod/pod.go [hunk 4]
```diff
@@ -666,11 +666,13 @@ func (p *Pod) AllContainersByID() ([]string, error) {
 func (p *Pod) AllContainers() ([]*Container, error) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
+	return p.allContainers()
+}
 
+func (p *Pod) allContainers() ([]*Container, error) {
 	if !p.valid {
```
Comment (mheon) [inline libpod/pod.go]: Move the valid check into AllContainers, we assume that whoever takes the lock first does the check

File: libpod/pod.go [hunk 5]
```diff
@@ -746,3 +748,38 @@ func (p *Pod) Inspect() (*PodInspect, error) {
 	}
 	return &inspectData, nil
 }
+
+type PodContainerStats struct {
```
Comment (mheon) [inline libpod/pod.go]: Lint is going to complain this has no comment

File: cmd/podman/pod_stats.go (hunk 1 above)
Comment (baude) [reply cmd/podman/pod_stats.go]: sure

File: cmd/podman/pod_stats.go (hunk 2 above)
Comment (baude) [reply cmd/podman/pod_stats.go]: done

File: cmd/podman/pod_stats.go (hunk 3 above)
Comment (baude) [reply cmd/podman/pod_stats.go]: nope

File: libpod/pod.go (hunk 4 above)
Comment (baude) [reply libpod/pod.go]: done

File: libpod/pod.go (hunk 5 above)
Comment (baude) [reply libpod/pod.go]: done

File: libpod/runtime_pod.go [hunk 6]
```diff
@@ -129,3 +130,59 @@ func (r *Runtime) GetLatestPod() (*Pod, error) {
...
+	containers, err := r.GetRunningContainers()
+	if err != nil {
+		return nil, err
+	}
+	// Assemble a slice of cids for lookup
+	for _, c := range containers {
+		cids = append(cids, c.ID())
+	}
+	for _, p := range pods {
+		containersByID, err := p.AllContainersByID()
```
Comment (mheon) [inline libpod/runtime_pod.go]: No need for this - go through all the containers from `GetRunningContainers` and build a set of `c.Pod()`, then get all of those pods

File: libpod/runtime_pod.go [hunk 7]
```diff
@@ -129,3 +130,59 @@ func (r *Runtime) GetLatestPod() (*Pod, error) {
...
+				break
+			}
+		}
+	}
+	return runningPods, nil
+}
+
+// GetPodsByList returns a slice of pods given a string list of wanted pod
+// ids or names
+func (r *Runtime) GetPodsByList(podList []string) ([]*Pod, error) {
```
Comment (mheon) [inline libpod/runtime_pod.go]: Either add a `GetContainersByList` or move this into cmd/podman - let's avoid adding things we don't have a container equivalent for.

File: libpod/runtime_pod.go (hunk 6 above)
Comment (baude) [reply libpod/runtime_pod.go]: why is that better

Comment (mheon) [reply libpod/runtime_pod.go]: You skip all the expensive `AllContainersByID` calls

Comment (baude) [reply libpod/runtime_pod.go]: fixed

File: libpod/runtime_pod.go (hunk 7 above)
Comment (baude) [reply libpod/runtime_pod.go]: done

**Comment by baude** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
Needs a rebase.  Sorry.

File: cmd/podman/pod_stats.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+			Usage: "disable streaming stats and only pull the first result, default setting is false",
+		},
+		cli.BoolFlag{
+			Name:  "no-reset",
+			Usage: "disable resetting the screen between intervals",
+		},
+		cli.StringFlag{
+			Name:  "format",
+			Usage: "pretty-print container statistics to JSON or using a Go template",
+		}, LatestFlag,
```
Comment (haircommander) [inline cmd/podman/pod_stats.go]: I'm still confused why LatestFlag is an option but not handled around line 90. if podman pod stats -l will just show all running pods, no?

Comment (haircommander) [inline cmd/podman/pod_stats.go]: This should also be LatestPodFlag if latest is being included

Comment (baude) [reply cmd/podman/pod_stats.go]: just not going to let that go are you ... :) my error, lemme fix

Comment (baude) [reply cmd/podman/pod_stats.go]: fixed

Comment (baude) [reply cmd/podman/pod_stats.go]: fixed

**Comment by haircommander** [comment]
LGTM

**Comment by baude** [comment]
merge me

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**